In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import roc_auc_score

# Santander Customer Transaction Prediction (KAGGLE)

Leer de que trata la competenciagit lfs install

https://www.kaggle.com/c/santander-customer-transaction-prediction/overview

# Bajada de Dataset

Crear la carpeta /data en el mismo directorio donde clonó el repositorio.

Guardar los archivos train.csv y test.csv en esta carpeta.

Los mismos se encuentran en el link mensionado arriba.

# Cargar datos de train

In [2]:
df_train = pd.read_csv('data/train.csv')

In [3]:
df_train

ID_code  target    var_0   var_1    var_2   var_3    var_4  \
0            train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607   
1            train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622   
2            train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825   
3            train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846   
4            train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772   
...              ...     ...      ...     ...      ...     ...      ...   
199995  train_199995       0  11.4880 -0.4956   8.2622  3.5142  10.3404   
199996  train_199996       0   4.9149 -2.4484  16.7052  6.6345   8.3096   
199997  train_199997       0  11.2232 -5.0518  10.5127  5.6456   9.3410   
199998  train_199998       0   9.7148 -8.6098  13.6104  5.7930  12.5173   
199999  train_199999       0  10.8762 -5.7105  12.1183  8.0328  11.5577   

          var_5   var_6    var_7  ...  var_190  var_191  var_192  var_193  \
0       -9.2834  5.1187  18.6266  ...   4.4354   3.9642   3.1364   1.6910   
1        7.0433  5.6208  16.5338  ...   7.6421   7.7214   2.5837  10.9516   
2       -9.0837  6.9427  14.6155  ...   2.9057   9.7905   1.6704   1.6858   
3       -1.8361  5.8428  14.9250  ...   4.4666   4.7433   0.7178   1.4214   
4        2.4486  5.9405  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942   
...         ...     ...      ...  ...      ...      ...      ...      ...   
199995  11.6081  5.6709  15.1516  ...   6.1415  13.2305   3.9901   0.9388   
199996 -10.5628  5.8802  21.5940  ...   4.9611   4.6549   0.6998   1.8341   
199997  -5.4086  4.5555  21.5571  ...   4.0651   5.4414   3.1032   4.8793   
199998   0.5339  6.0479  17.0152  ...   2.6840   8.6587   2.7337  11.1178   
199999   0.3488  5.2839  15.2058  ...   8.9842   1.6893   0.1276   0.3766   

        var_194  var_195  var_196  var_197  var_198  var_199  
0       18.5227  -2.3978   7.8784   8.5635  12.7803  -1.0914  
1       15.4305   2.0339   8.1267   8.7889  18.3560   1.9518  
2       21.6042   3.1417  -6.5213   8.2675  14.7222   0.3965  
3       23.0347  -1.2706  -2.9275  10.2922  17.9697  -8.9996  
4       13.2876  -1.5121   3.9267   9.5031  17.9974  -8.8104  
...         ...      ...      ...      ...      ...      ...  
199995  18.0249  -1.7939   2.1661   8.5326  16.6660 -17.8661  
199996  22.2717   1.7337  -2.1651   6.7419  15.9054   0.3388  
199997  23.5311  -1.5736   1.2832   8.7155  13.8329   4.1995  
199998  20.4158  -0.0786   6.7980  10.0342  15.5289 -13.9001  
199999  15.2101  -2.4907  -2.2342   8.1857  12.1284   0.1385  

[200000 rows x 202 columns]

In [4]:
# Miramos las primeras 5 observaciones del dataset
df_train.head()

ID_code  target    var_0   var_1    var_2   var_3    var_4   var_5   var_6  \
0  train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607 -9.2834  5.1187   
1  train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622  7.0433  5.6208   
2  train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825 -9.0837  6.9427   
3  train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846 -1.8361  5.8428   
4  train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772  2.4486  5.9405   

     var_7  ...  var_190  var_191  var_192  var_193  var_194  var_195  \
0  18.6266  ...   4.4354   3.9642   3.1364   1.6910  18.5227  -2.3978   
1  16.5338  ...   7.6421   7.7214   2.5837  10.9516  15.4305   2.0339   
2  14.6155  ...   2.9057   9.7905   1.6704   1.6858  21.6042   3.1417   
3  14.9250  ...   4.4666   4.7433   0.7178   1.4214  23.0347  -1.2706   
4  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942  13.2876  -1.5121   

   var_196  var_197  var_198  var_199  
0   7.8784   8.5635  12.7803  -1.0914  
1   8.1267   8.7889  18.3560   1.9518  
2  -6.5213   8.2675  14.7222   0.3965  
3  -2.9275  10.2922  17.9697  -8.9996  
4   3.9267   9.5031  17.9974  -8.8104  

[5 rows x 202 columns]

In [5]:
df_train.columns

Index(['ID_code', 'target', 'var_0', 'var_1', 'var_2', 'var_3', 'var_4',
       'var_5', 'var_6', 'var_7',
       ...
       'var_190', 'var_191', 'var_192', 'var_193', 'var_194', 'var_195',
       'var_196', 'var_197', 'var_198', 'var_199'],
      dtype='object', length=202)

In [6]:
df_train.describe()

target          var_0          var_1          var_2  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        0.100490      10.679914      -1.627622      10.715192   
std         0.300653       3.040051       4.050044       2.640894   
min         0.000000       0.408400     -15.043400       2.117100   
25%         0.000000       8.453850      -4.740025       8.722475   
50%         0.000000      10.524750      -1.608050      10.580000   
75%         0.000000      12.758200       1.358625      12.516700   
max         1.000000      20.315000      10.376800      19.353000   

               var_3          var_4          var_5          var_6  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        6.796529      11.078333      -5.065317       5.408949   
std         2.043319       1.623150       7.863267       0.866607   
min        -0.040200       5.074800     -32.562600       2.347300   
25%         5.254075       9.883175     -11.200350       4.767700   
50%         6.825000      11.108250      -4.833150       5.385100   
75%         8.324100      12.261125       0.924800       6.003000   
max        13.188300      16.671400      17.251600       8.447700   

               var_7          var_8  ...        var_190        var_191  \
count  200000.000000  200000.000000  ...  200000.000000  200000.000000   
mean       16.545850       0.284162  ...       3.234440       7.438408   
std         3.418076       3.332634  ...       4.559922       3.023272   
min         5.349700     -10.505500  ...     -14.093300      -2.691700   
25%        13.943800      -2.317800  ...      -0.058825       5.157400   
50%        16.456800       0.393700  ...       3.203600       7.347750   
75%        19.102900       2.937900  ...       6.406200       9.512525   
max        27.691800      10.151300  ...      18.440900      16.716500   

             var_192        var_193        var_194        var_195  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        1.927839       3.331774      17.993784      -0.142088   
std         1.478423       3.992030       3.135162       1.429372   
min        -3.814500     -11.783400       8.694400      -5.261000   
25%         0.889775       0.584600      15.629800      -1.170700   
50%         1.901300       3.396350      17.957950      -0.172700   
75%         2.949500       6.205800      20.396525       0.829600   
max         8.402400      18.281800      27.928800       4.272900   

             var_196        var_197        var_198        var_199  
count  200000.000000  200000.000000  200000.000000  200000.000000  
mean        2.303335       8.908158      15.870720      -3.326537  
std         5.454369       0.921625       3.010945      10.438015  
min       -14.209600       5.960600       6.299300     -38.852800  
25%        -1.946925       8.252800      13.829700     -11.208475  
50%         2.408900       8.888200      15.934050      -2.819550  
75%         6.556725       9.593300      18.064725       4.836800  
max        18.321500      12.000400      26.079100      28.500700  

[8 rows x 201 columns]

In [7]:
df_train.shape

(200000, 202)

### Pregunta 1
#### ¿Cual es la dimensión de los features?
##### 200
#### ¿Cuantas clases hay?
##### 2
#### ¿Cuantas observaciones hay en train?
##### 200000

# Cargar datos de test

In [8]:
# TODO: Importar datos de test y asignarlos a la variable df_test desde la carpeta data/
df_test  = pd.read_csv('data/test.csv')

In [9]:
# TODO: Inspeccionar test
# Puede ejecutar los métodos que crea necesarios. Recomendamos describe, columns, shape, head

In [10]:
df_test

ID_code    var_0    var_1    var_2   var_3    var_4    var_5  \
0            test_0  11.0656   7.7798  12.9536  9.4292  11.4327  -2.3805   
1            test_1   8.5304   1.2543  11.3047  5.1858   9.1974  -4.0117   
2            test_2   5.4827 -10.3581  10.1407  7.0479  10.2628   9.8052   
3            test_3   8.5374  -1.3222  12.0220  6.5749   8.8458   3.1744   
4            test_4  11.7058  -0.1327  14.1295  7.7506   9.1035  -8.5848   
...             ...      ...      ...      ...     ...      ...      ...   
199995  test_199995  13.1678   1.0136  10.4333  6.7997   8.5974  -4.1641   
199996  test_199996   9.7171  -9.1462   7.3443  9.1421  12.8936   3.0191   
199997  test_199997  11.6360   2.2769  11.2074  7.7649  12.6796  11.3224   
199998  test_199998  13.5745  -0.5134  13.6584  7.4855  11.2241 -11.3037   
199999  test_199999  10.4664   1.8070  10.2277  6.0654  10.0258   1.0789   

         var_6    var_7   var_8  ...  var_190  var_191  var_192  var_193  \
0       5.8493  18.2675  2.1337  ...  -2.1556  11.8495  -1.4300   2.4508   
1       6.0196  18.6316 -4.4131  ...  10.6165   8.8349   0.9403  10.1282   
2       4.8950  20.2537  1.5233  ...  -0.7484  10.9935   1.9803   2.1800   
3       4.9397  20.5660  3.3755  ...   9.5702   9.0766   1.6580   3.5813   
4       6.8595  10.6048  2.9890  ...   4.2259   9.1723   1.2835   3.3778   
...        ...      ...     ...  ...      ...      ...      ...      ...   
199995  4.8579  14.7625 -2.7239  ...   2.0544   9.6849   4.6734  -1.3660   
199996  5.6888  18.8862  5.0915  ...   5.0071   6.6548   1.8197   2.4104   
199997  5.3883  18.3794  1.6603  ...   5.1536   2.6498   2.4937  -0.0637   
199998  4.1959  16.8280  5.3208  ...   3.4259   8.5012   2.2713   5.7621   
199999  4.8879  14.4892 -0.5902  ...   0.1398   9.2828   1.3601   4.8985   

        var_194  var_195  var_196  var_197  var_198  var_199  
0       13.7112   2.4669   4.3654  10.7200  15.4722  -8.7197  
1       15.5765   0.4773  -1.4852   9.8714  19.1293 -20.9760  
2       12.9813   2.1281  -7.1086   7.0618  19.8956 -23.1794  
3       15.1874   3.1656   3.9567   9.2295  13.0168  -4.2108  
4       19.5542  -0.2860  -5.1612   7.2882  13.9260  -9.1846  
...         ...      ...      ...      ...      ...      ...  
199995  12.8721   1.2013  -4.6195   9.1568  18.2102   4.8801  
199996  18.9037  -0.9337   2.9995   9.1112  18.1740 -20.7689  
199997  20.0609  -1.1742  -4.1524   9.1933  11.7905 -22.2762  
199998  17.0056   1.1763  -2.3761   8.1079   8.7735  -0.2122  
199999  20.0926  -1.3048  -2.5981  10.3378  14.3340  -7.7094  

[200000 rows x 201 columns]

In [11]:
df_test.head()

ID_code    var_0    var_1    var_2   var_3    var_4   var_5   var_6  \
0  test_0  11.0656   7.7798  12.9536  9.4292  11.4327 -2.3805  5.8493   
1  test_1   8.5304   1.2543  11.3047  5.1858   9.1974 -4.0117  6.0196   
2  test_2   5.4827 -10.3581  10.1407  7.0479  10.2628  9.8052  4.8950   
3  test_3   8.5374  -1.3222  12.0220  6.5749   8.8458  3.1744  4.9397   
4  test_4  11.7058  -0.1327  14.1295  7.7506   9.1035 -8.5848  6.8595   

     var_7   var_8  ...  var_190  var_191  var_192  var_193  var_194  var_195  \
0  18.2675  2.1337  ...  -2.1556  11.8495  -1.4300   2.4508  13.7112   2.4669   
1  18.6316 -4.4131  ...  10.6165   8.8349   0.9403  10.1282  15.5765   0.4773   
2  20.2537  1.5233  ...  -0.7484  10.9935   1.9803   2.1800  12.9813   2.1281   
3  20.5660  3.3755  ...   9.5702   9.0766   1.6580   3.5813  15.1874   3.1656   
4  10.6048  2.9890  ...   4.2259   9.1723   1.2835   3.3778  19.5542  -0.2860   

   var_196  var_197  var_198  var_199  
0   4.3654  10.7200  15.4722  -8.7197  
1  -1.4852   9.8714  19.1293 -20.9760  
2  -7.1086   7.0618  19.8956 -23.1794  
3   3.9567   9.2295  13.0168  -4.2108  
4  -5.1612   7.2882  13.9260  -9.1846  

[5 rows x 201 columns]

In [12]:
df_test.columns

Index(['ID_code', 'var_0', 'var_1', 'var_2', 'var_3', 'var_4', 'var_5',
       'var_6', 'var_7', 'var_8',
       ...
       'var_190', 'var_191', 'var_192', 'var_193', 'var_194', 'var_195',
       'var_196', 'var_197', 'var_198', 'var_199'],
      dtype='object', length=201)

In [13]:
df_test.describe()

var_0          var_1          var_2          var_3  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean       10.658737      -1.624244      10.707452       6.788214   
std         3.036716       4.040509       2.633888       2.052724   
min         0.188700     -15.043400       2.355200      -0.022400   
25%         8.442975      -4.700125       8.735600       5.230500   
50%        10.513800      -1.590500      10.560700       6.822350   
75%        12.739600       1.343400      12.495025       8.327600   
max        22.323400       9.385100      18.714100      13.142000   

               var_4          var_5          var_6          var_7  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean       11.076399      -5.050558       5.415164      16.529143   
std         1.616456       7.869293       0.864686       3.424482   
min         5.484400     -27.767000       2.216400       5.713700   
25%         9.891075     -11.201400       4.772600      13.933900   
50%        11.099750      -4.834100       5.391600      16.422700   
75%        12.253400       0.942575       6.005800      19.094550   
max        16.037100      17.253700       8.302500      28.292800   

               var_8          var_9  ...        var_190        var_191  \
count  200000.000000  200000.000000  ...  200000.000000  200000.000000   
mean        0.277135       7.569407  ...       3.189766       7.458269   
std         3.333375       1.231865  ...       4.551239       3.025189   
min        -9.956000       4.243300  ...     -14.093300      -2.407000   
25%        -2.303900       6.623800  ...      -0.095000       5.166500   
50%         0.372000       7.632000  ...       3.162400       7.379000   
75%         2.930025       8.584825  ...       6.336475       9.531100   
max         9.665500      11.003600  ...      20.359000      16.716500   

             var_192        var_193        var_194        var_195  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        1.925944       3.322016      17.996967      -0.133657   
std         1.479966       3.995599       3.140652       1.429678   
min        -3.340900     -11.413100       9.382800      -4.911900   
25%         0.882975       0.587600      15.634775      -1.160700   
50%         1.892600       3.428500      17.977600      -0.162000   
75%         2.956000       6.174200      20.391725       0.837900   
max         8.005000      17.632600      27.947800       4.545400   

             var_196        var_197        var_198        var_199  
count  200000.000000  200000.000000  200000.000000  200000.000000  
mean        2.290899       8.912428      15.869184      -3.246342  
std         5.446346       0.920904       3.008717      10.398589  
min       -13.944200       6.169600       6.584000     -39.457800  
25%        -1.948600       8.260075      13.847275     -11.124000  
50%         2.403600       8.892800      15.943400      -2.725950  
75%         6.519800       9.595900      18.045200       4.935400  
max        15.920700      12.275800      26.538400      27.907400  

[8 rows x 200 columns]

In [14]:
df_test.shape

(200000, 201)

### Pregunta 2
#### ¿Cuantas observaciones hay en test?
##### 200000
#### ¿Que columna está en train pero no en test?
##### target

### Pregunta 3
#### Cual es el porcentaje de 0s que hay en este dataset?

In [15]:
(df_train['target'].value_counts( normalize=True))*100

0    89.951
1    10.049
Name: target, dtype: float64

#### Respuesta:
##### 89.951 %

### Pregunta 4
#### ¿Qué accuracy obtendría si independiente de la entrada mi modelo siempre predigiera un 0? 
##### 0.89951
#### ¿Qué area debajo de la curva ROC obtendría si independiente de la entrada mi modelo siempre predigiera un 0?
##### 0.5

# Dividir data en Train y Val

In [16]:
N_train = 160_000
X_train = df_train.drop(columns=['ID_code', 'target']).values[:N_train]
X_val = df_train.drop(columns=['ID_code', 'target']).values[N_train:]
y_train = df_train['target'].values[:N_train]
y_val = df_train['target'].values[N_train:]

# Regresión logística

In [17]:
from sklearn.linear_model import LogisticRegression

In [18]:
model = LogisticRegression(solver='lbfgs', max_iter=100)

In [19]:
%time model.fit(X_train, y_train)

Wall time: 6.1 s


C:\Users\USUARIO\anaconda3\envs\DeepLeaning38\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

### Pregunta 5
#### Cuales de las siguientes afirmaciones son ciertas:
#### *El solver no terminó de converger
#### *Una posible solución es aumentar la cantidad de iteraciones
##### Ambas son ciertas

### Pregunta 6
#### Para los datos de Training:

In [20]:
acc = model.score(X_train, y_train)
auc = roc_auc_score(y_train, model.predict_proba(X_train)[:,1])
print(auc)
print(acc)

0.8506506508377423
0.9120875


#### ¿Cuánto es el accuracy con max_iter=100 a pesar de que da el warning?
##### 0.9120875
#### ¿Cuánto da el auc ROC?
##### 0.8506506508377423 

### Pregunta 7
#### Para los datos de Validation:

In [21]:
print(roc_auc_score(y_val, model.predict_proba(X_val)[:,1]))
print(model.score(X_val, y_val))

0.8527900125633393
0.912175


#### ¿Cuánto es el accuracy con max_iter=100 a pesar de que da el warning?
##### 0.912175
#### ¿Cuánto da el auc ROC?
##### 0.8527900125633393

### Pregunta 8
#### Cuales de las afirmaciones son correctas:
#### *No hay overfitting, es decir, el modelo no memorizó los datos
#### *Tanto el accuracy como el ROC mejoraron respecto al "modelo" que predice siempre 0
#### *Para calcular el auc ROC le paso al método las predicciones de probabilidad que dio el modelo
##### Todas son correctas

### Problema 9
#### ¿Cuántos parámetros aprendió el modelo?
#### fórmula: (D+1) , #D=Features de entrada

In [41]:
(X_train).shape

(160000, 200)

#### Respuesta:
##### 201

# Normalizar los datos

In [22]:
means = X_train.mean(axis=0)
stds = X_train.std(axis=0)

In [23]:
X_train_normalized = (X_train - means)/stds
X_val_normalized = (X_val - means)/stds

In [24]:
model_2 = LogisticRegression(solver='lbfgs')
%time model_2.fit(X_train_normalized, y_train)

Wall time: 909 ms


LogisticRegression()

In [25]:
print(model_2.score(X_train_normalized, y_train))
print(roc_auc_score(y_train, model_2.predict_proba(X_train_normalized)[:,1]))

0.91440625
0.8606427353156717


In [26]:
print(model_2.score(X_val_normalized, y_val))
print(roc_auc_score(y_val, model_2.predict_proba(X_val_normalized)[:,1]))

0.914525
0.8629075884886785


### Pregunta 10
#### Para resolver el problema de la convergencia vamos entonces a normalizar los datos de entrada. Calcule la media y el desvio standard de cada columna. Reste la media y divida por el desvio a cada columna y vuelva a hacer el fit del modelo.
#### Indicar cuales son vedaderas:
#### *El fit tarda menos luego de normalizar sin dar warning
#### *Tanto el accuracy como el auc ROC mejoran levemente 
##### Ambas son verdaderas.

### Pregunta 11
#### ¿Cuáles son los dos parámetros que contribuyen más en la regresión logística para hacer la predicción?

In [27]:
# Ordenamos el valor absoluto de los parámetros de mayor a menor
# Coeficientes ordenados en función de su de mayor importancia
np.argsort(np.abs(model_2.coef_))[0][::-1]

array([ 81, 139,   6,  12,  53, 110,  76,  99, 146, 174,  26,  21,  22,
        80, 190, 166, 165, 198,  13,  34, 169,  44,   2, 148,   0,  40,
       133, 179, 170,  78,  94,   1, 184, 109,  33, 115,  92,  67, 149,
       108, 191, 122, 173, 154,  18,  86, 192, 118, 107, 121, 147,  95,
         9,  75,  35, 164, 177, 197, 172,  36, 127,  89, 123, 155,  91,
       188,  56,  87,  71,  48,   5,  93, 162, 106, 157, 130, 141, 145,
        24, 151,  32, 167, 163, 150, 186, 119,  49,  31, 180,  23, 111,
       195,  90, 131, 125, 137, 114, 199,  43, 116, 135,  52,  58, 128,
        70, 104, 175, 112, 132, 105,  11, 196,  85,  82, 194,  51,  28,
       142,  83,  66, 134, 144, 138,  74,  45, 156,  77,  55,  97, 140,
        20,  54, 193,  57,  88, 178,   8, 102, 113,  62,  15, 143, 159,
       187, 181, 171,  63,  72,  64,  50,  59, 120, 168, 182, 101,  25,
        68,   3,  65, 152,   4,  84,  37,  42, 176,  61, 129,  19, 189,
        69,  47,  60,  16,  27,  29,  79,  73, 158,  96, 160,  4

##### Respuesta:
##### 81, 139

### Pregunta 12
#### ¿Cuáles son los dos parámetros que más contribuyen a predecir un cero?

In [28]:
# Ordenados de menor a mayor
np.argsort(model_2.coef_)[0]

array([ 81, 139,  12,  76, 146, 174,  21,  80, 166, 165, 198,  13,  34,
       169,  44, 148, 109,  33, 115,  92, 149, 108, 122, 154,  86, 192,
       107, 121,   9,  75, 177, 197, 172,  36, 127, 123, 188,  56,  87,
        93, 141, 150, 186,  31,  23, 131, 114,  43, 116,  58, 104, 132,
        85, 194,  28, 142,  83,  45, 156,  77,  20,  54, 193,  57,  88,
       178, 102, 113, 143,  63,  72,  64,  50,  59, 120, 182, 101,  68,
       152,  42, 129,  27,  73, 158, 160,  98,  14, 153,  39, 136, 183,
        41, 103,  10,   7, 185,  38,  17,  30, 161, 100, 117, 126, 124,
        46,  96,  79,  29,  16,  60,  47,  69, 189,  19,  61, 176,  37,
        84,   4,  65,   3,  25, 168, 171, 181, 187, 159,  15,  62,   8,
       140,  97,  55,  74, 138, 144, 134,  66,  51,  82, 196,  11, 105,
       112, 175,  70, 128,  52, 135, 199, 137, 125,  90, 195, 111, 180,
        49, 119, 163, 167,  32, 151,  24, 145, 130, 157, 106, 162,   5,
        48,  71,  91, 155,  89, 164,  35,  95, 147, 118,  18, 17

##### Respuesta:
##### 81, 139

### Pregunta 13
#### ¿Cuáles son los dos parámetros que más contribuyen a predecir un uno?

In [29]:
# Ordenados de mayor a menor
np.argsort(model_2.coef_)[0][::-1]

array([  6,  53, 110,  99,  26,  22, 190,   2,   0,  40, 133, 179, 170,
        78,  94,   1, 184,  67, 191, 173,  18, 118, 147,  95,  35, 164,
        89, 155,  91,  71,  48,   5, 162, 106, 157, 130, 145,  24, 151,
        32, 167, 163, 119,  49, 180, 111, 195,  90, 125, 137, 199, 135,
        52, 128,  70, 175, 112, 105,  11, 196,  82,  51,  66, 134, 144,
       138,  74,  55,  97, 140,   8,  62,  15, 159, 187, 181, 171, 168,
        25,   3,  65,   4,  84,  37, 176,  61,  19, 189,  69,  47,  60,
        16,  29,  79,  96,  46, 124, 126, 117, 100, 161,  30,  17,  38,
       185,   7,  10, 103,  41, 183, 136,  39, 153,  14,  98, 160, 158,
        73,  27, 129,  42, 152,  68, 101, 182, 120,  59,  50,  64,  72,
        63, 143, 113, 102, 178,  88,  57, 193,  54,  20,  77, 156,  45,
        83, 142,  28, 194,  85, 132, 104,  58, 116,  43, 114, 131,  23,
        31, 186, 150, 141,  93,  87,  56, 188, 123, 127,  36, 172, 197,
       177,  75,   9, 121, 107, 192,  86, 154, 122, 108, 149,  9

##### Respuesta:
##### 6,  53

### Pregunta 14

#### ¿Cuáles son los 3 parámetros tienen mayor correlación con la salida (No importa el signo sino el valor absoluto)?

In [30]:
# Calculamos correlación de cada variable con la salida
correlations = []
for i in range(200):
    correlations.append(np.corrcoef(y_train, X_train_normalized[:,i])[1,0])

In [31]:
np.argsort(np.abs(correlations))[::-1]

array([ 81, 139,  12, 110,   6,  26,  53, 146, 174,  76,  99,  80, 166,
        22, 165,  21, 190, 148,   2,  13,  34, 133,   0, 198, 169, 109,
       179,  44,  40,   1, 115, 149, 170, 184,  78,  94,  92, 108,  67,
       191,  33,  18, 154, 173,  86, 122, 192, 147, 118,   9, 121,  75,
        95, 164, 123,  35,  91,  36, 172, 127, 107, 155,  89, 177, 197,
        56,  93, 188,  87,  71, 162,  48, 157, 106, 141, 145,   5,  32,
       163, 131, 167, 186, 119,  49,  24, 151, 130,  90, 111, 135, 180,
       125,  43,  51, 114, 195, 199,  52,  23, 150,  31, 116,  85,  70,
       137, 105, 128, 104,  58, 112, 196, 132,  11,  28, 175,  66,  82,
       194, 156,  83, 142,  45, 144,  88,  74, 178, 138,  77,   8,  97,
       134,  20,  55, 193,  54,  15, 102,  57, 159, 140, 171,  62, 113,
       187,  63, 143,  64,  25, 181,  50, 120, 168,  59,   3,  72,  65,
        68,  84, 101,  19,   4,  37,  61,  69, 189, 152, 182,  42,  16,
       129, 176,  47,  79, 153,  14, 183,  46, 160,  73,  60,  9

##### Respuesta:
##### 81, 139,  12

### Pregunta 15
#### Agregamos 200 features más que son los cuadrados de cada feature original.

# Feature engineering

In [32]:
# Agregamos 200 columnas con el cuadrado de cada columna normaliado a train
X_train_sq = X_train_normalized**2
means_sq = X_train_sq.mean(axis=0)
stds_sq = X_train_sq.std(axis=0)
X_train_FE = np.append(X_train_normalized, (X_train_sq - means_sq)/stds_sq, axis=1)

# Agregamos el cuadrado a validación normalizando con la media y std obtenida en train
X_val_sq = X_val_normalized**2
X_val_FE = np.append(X_val_normalized, (X_val_normalized**2-means_sq)/stds_sq, axis=1)

In [33]:
model_3 = LogisticRegression(solver='lbfgs')
%time model_3.fit(X_train_FE, y_train)

Wall time: 1.74 s


LogisticRegression()

In [34]:
print(model_3.score(X_train_FE, y_train))
print(roc_auc_score(y_train, model_3.predict_proba(X_train_FE)[:,1]))

0.9218625
0.8910807326950416


### Para train:

#### ¿Cuánto da el accuracy ahora?
##### 0.9218625
#### ¿Cuánto da el auc ROC?
##### 0.8910807326950416

In [ ]:
# Calculamos Accuray y Auc_ROC para Validation

In [35]:
print(model_3.score(X_val_FE, y_val))
print(roc_auc_score(y_val, model_3.predict_proba(X_val_FE)[:,1]))

0.9221
0.8927208962331821


In [ ]:
# Calculamos La Cross Entropy 

In [36]:
from sklearn.metrics import log_loss

In [37]:
Loss_train = log_loss(y_train , model.predict_proba(X_train)[:,1])
print(Loss_train)

0.23784959773845052


In [38]:
Loss_val = log_loss(y_val , model.predict_proba(X_val)[:,1])
print(Loss_val)

0.2378570360097157


### Pregunta 16
#### ¿Cuántos parámetros aprendió el modelo?

In [40]:
(X_train_FE).shape

(160000, 400)

##### Fórmula: D+1 , #D=400 features de entrada
##### 401

### Pregunta 17
#### Indicar cuales son verdaderas
#### *El accuracy tenía menos margen para aumentar ya que estaba en 0.914 contra el auc ROC que estaba en el orden de 0.863. No sería preciso hacer esta comparación tan directa.
#### *El auc ROC mide la capacidad de un modelo de separar los datos
#### *Cuando la entropía cruzada tiende a cero, el auc ROC debería tender al 100%
#### *Podríamos tener un caso donde la entropía cruzada no sea cero pero el accuracy sea de 100%
##### Todas son verdaderas.